app/src/main/java/org/simple/clinic/scheduleappointment/ScheduleAppointmentSheetController.kt

```diff
@@ -0,0 +1,99 @@
+
+  override fun apply(upstream: Observable<UiEvent>): Observable<UiChange> {
+    val replayedEvents = upstream.compose(ReportAnalyticsEvents()).replay().refCount()
+
+    return Observable.mergeArray(
+        sheetCreates(replayedEvents),
+        dateIncrements(replayedEvents),
+        dateDecrements(replayedEvents),
+        schedulingSkips(replayedEvents),
+        scheduleCreates(replayedEvents))
+  }
+
+  private fun sheetCreates(events: Observable<UiEvent>): Observable<UiChange> {
+    return events.ofType<SheetCreated>()
+        .map { { ui: Ui -> ui.updateDisplayedDate(it.initialIndex) } }
+  }
+
+  private fun dateIncrements(events: Observable<UiEvent>): Observable<UiChange> {
+    val lastPositionStream = events.ofType<IncrementAppointmentDate>()
+        .filter { (current, size) -> isSecondLastItem(current, size) }
+        .map { { ui: Ui -> ui.enableIncrementButton(false) } }
+
+    val notLastPositionStream = events.ofType<IncrementAppointmentDate>()
+        .filter { (current, size) -> isNotLastItem(current, size) }
+
+    val updateDateStream = notLastPositionStream
+        .map { { ui: Ui -> ui.updateDisplayedDate(newIndex = it.currentIndex.inc()) } }
+
+    val enableButtonStream = notLastPositionStream
+        .distinctUntilChanged()
+        .map { { ui: Ui -> ui.enableIncrementAndDecrementButtons() } }
+
+    return Observable.merge(lastPositionStream, updateDateStream, enableButtonStream)
+  }
+
+  private fun isSecondLastItem(current: Int, size: Int) = current == size - 2
+
+  private fun isNotLastItem(current: Int, size: Int) = current != size - 1
+
+  private fun dateDecrements(events: Observable<UiEvent>): Observable<UiChange> {
+    val notFirstPositionStream = events.ofType<DecrementAppointmentDate>()
+        .filter { it.currentIndex != 0 }
+
+    val updateDateStream = notFirstPositionStream
+        .map { { ui: Ui -> ui.updateDisplayedDate(newIndex = it.currentIndex.dec()) } }
+
+    val enableButtonStream = notFirstPositionStream
+        .distinctUntilChanged()
+        .map { { ui: Ui -> ui.enableIncrementAndDecrementButtons() } }
+
+    val firstPositionStream = events.ofType<DecrementAppointmentDate>()
+        .filter { it.currentIndex == 1 }
+        .map { { ui: Ui -> ui.enableDecrementButton(false) } }
+
+    return Observable.merge(updateDateStream, enableButtonStream, firstPositionStream)
+  }
+
+  private fun schedulingSkips(events: Observable<UiEvent>): Observable<UiChange> {
+    return events.ofType<SkipScheduling>()
+        .map { { ui: Ui -> ui.closeSheet() } }
+  }
+
+  private fun scheduleCreates(events: Observable<UiEvent>): Observable<UiChange> {
+    val patientUuidStream = events.ofType<SheetCreated>()
+        .map { it.patientUuid }
+
+    return events.ofType<ScheduleAppointment>()
+        .withLatestFrom(patientUuidStream)
+        .map { (scheduleEvent, patientUuid) -> patientUuid to appointmentLocalDate(scheduleEvent.selectedDateState.second) }
```

Comment:
Why have such long lines da.

This can be replaced with:

```kotlin
events.ofType<ScheduleAppointment>()
  .map { appointmentLocalDate(it.selectedDateState.second) }
  .withLatestFrom(patientUuidStream)
```